### PR TITLE
feature: support create a container with a given id

### LINKED ISSUE #2180: [proposal] create a container with a given ID
### Ⅰ. Issue Description
Container ID is generated by pouch automatically. I have a management tool in which container ID is related to other members. When restoring a rich container in fault recover scene，I want to create a container with a given ID (actually the ID is the same as the container which is in fault) . It will be nice that `pouch run` can support a given container ID

### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1.
2.
3.

### Ⅵ. Environment:

- pouch version (use `pouch version`):
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by wangforthinker**
Signed-off-by: allen.wang <allen.wq@alipay.com>

### Ⅰ. Describe what this PR did
it can create container with a  given id.

### Ⅱ. Does this pull request fix one issue?

fixes #2180 

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)
add test TestBasicWithSpecificID and TestInvalidSpecificID in api_container_create_test.go.

### Ⅳ. Describe how to verify it
verify it by test.

### Ⅴ. Special notes for reviews
none.

### CODE REVIEW

**Comment by allencloud** [comment]
Could you please first to fix the conflict there in this pull request? @wangforthinker

**Comment by allencloud** [comment]
Could you please rebase to the latest master? @wangforthinker 
Since I found that in you commit have lots of changes from previous commits in the master.

**Comment by wangforthinker** [comment]
I have rebased it ok. @fuweid

**Comment by allencloud** [comment]
How about making the length of specificID fixed to be the same as the original generated length? @wangforthinker

**Comment by wangforthinker** [comment]
> How about making the length of specificID fixed to be the same as the original generated length? @wangforthinker

Good Idea! Now Set limit of specificID length to 64.

File: apis/swagger.yml [hunk 1]
```diff
@@ -2211,6 +2211,13 @@ definitions:
         description: "net priority."
         type: "integer"
         default: 0
+      SpecificID:
+        type: "string"
+        description: |
+          Create container with given id.
+          The length of given id should be 64.
+          The characters of given id should be in 0123456789abcdef.
+          By default, given id is unnecessary.
```
Comment (allencloud) [inline apis/swagger.yml]: Could you add the length here to help illustrate this?
```
minLength: 64
maxLength: 64
```

File: daemon/mgr/container.go [hunk 2]
```diff
@@ -293,10 +296,36 @@ func (mgr *ContainerManager) Create(ctx context.Context, name string, config *ty
 		return nil, errors.Wrapf(errtypes.ErrInvalidParam, "NetworkingConfig cannot be empty")
 	}
 
-	id, err := mgr.generateID()
-	if err != nil {
-		return nil, err
+	if config.SpecificID != "" {
```
Comment (allencloud) [inline daemon/mgr/container.go]: I think the function `Create` function is so long, could you please put the ID generation part encapsulated into a single function? @wangforthinker

**Comment by allencloud** [comment]
Please do not git commit file `registry_service_config.go`, since this file's generation would bring other issues. Thanks a lot. @wangforthinker

File: apis/server/container_bridge.go [hunk 3]
```diff
@@ -43,6 +43,11 @@ func (s *Server) createContainer(ctx context.Context, rw http.ResponseWriter, re
 	logCreateOptions("container", config)
 
 	name := req.FormValue("name")
+	//consider set specific id by url params
+	specificID := req.FormValue("recover-id")
```
Comment (allencloud) [inline apis/server/container_bridge.go]: Where to define `recover-id`?
Actually I do not find anywhere to define this. @wangforthinker

File: test/api_container_create_test.go [hunk 4]
```diff
@@ -279,3 +280,86 @@ func (suite *APIContainerCreateSuite) TestCreateWithMacAddress(c *check.C) {
 
 	DelContainerForceMultyTime(c, cname)
 }
+
+func (suite *APIContainerCreateSuite) TestBasicWithSpecificID(c *check.C) {
+	{
```
Comment (zhuangqh) [inline test/api_container_create_test.go]: missing test case. specificID is conflict with an existing container's ID

Comment (wangforthinker) [reply test/api_container_create_test.go]: add test for it!

File: daemon/mgr/container.go [hunk 5]
```diff
@@ -1869,3 +1877,24 @@ func (mgr *ContainerManager) execProcessGC() {
...
+		for _, c := range []byte(specificID) {
+			if !((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f')) {
+				return "", errors.Wrap(errtypes.ErrInvalidParam, "The characters of container id should be in '0123456789abcdef'")
+			}
+		}
+
+		if mgr.cache.Get(specificID).Exist() {
+			return "", errors.Wrap(errtypes.ErrAlreadyExisted, "container id: "+specificID)
+		}
+		return specificID, nil
```
[the remark this answers is not in the document]
Comment (wangforthinker) [reply daemon/mgr/container.go]: this is also not bad!

File: daemon/mgr/container.go (hunk 2 above)
Comment (wangforthinker) [reply daemon/mgr/container.go]: Have Done it!

File: apis/server/container_bridge.go (hunk 3 above)
Comment (wangforthinker) [reply apis/server/container_bridge.go]: Updated it to param specificID.

File: apis/swagger.yml (hunk 1 above)
Comment (wangforthinker) [reply apis/swagger.yml]: Have fixed it!

**Comment by allencloud** [comment]
LGTM

**Pull request merged**